### DISCUSSION TITLE
ColumnCollection is not a Mapping

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#9375] [Ideas]
### Describe the typing issue

ColumnCollection appears to inherit from Generic, but I suspect that it should be inheriting from Mapping.

When annotating a function in a generic way, so it shouldn't care too much about sqlalchemy internals, you get an error.

### To Reproduce

```python
def create_filters(columns: Mapping[Any, Any]): ...

create_filters(self._table.c)
```

### Error

```
aiohttp_admin/backends/sqlalchemy.py:73:53: error: Argument 1 to "create_filters" has incompatible type "ReadOnlyColumnCollection[str, Column[Any]]"; expected "Mapping[Any, Any]"  [arg-type]
                    query = query.where(*create_filters(self._table.c, filters))
                                                        ^~~~~~~~~~~~~
```

I only used `Any` here to verify that it was the `Mapping` itself causing the problem.

### Versions

- Python: 3.9
- SQLAlchemy: 2.0.4
- Type checker: mypy 1.0.1

**Comment by CaselIT** [comment]
Hi,

Thanks for reporting.
Yes, I guess that column collection can be considered a mapping, since it has keys, values and items like one.
Shame that mapping does not behave lake a Protocol here.

@zzzeek do you think we can just extend from it or should we just make type checkers believe we do?

**Comment by zzzeek** [comment]
I'm not sure extending from Mapping works because Mapping has a lot of methods, some of which we dont have, others which we do but dont do in a Mapping like way, in particular iteration.

```py
>>> t = table('t', column('a'), column('b'), column('c'))
>>> t.c.keys()
['a', 'b', 'c']
>>> list(t.c)
[<sqlalchemy.sql.elements.ColumnClause at 0x7f645dd7ceb0; a>, <sqlalchemy.sql.elements.ColumnClause at 0x7f645dec3310; b>, <sqlalchemy.sql.elements.ColumnClause at 0x7f645dd08df0; c>]
>>> t.c.values()
[<sqlalchemy.sql.elements.ColumnClause at 0x7f645dd7ceb0; a>, <sqlalchemy.sql.elements.ColumnClause at 0x7f645dec3310; b>, <sqlalchemy.sql.elements.ColumnClause at 0x7f645dd08df0; c>]

```

we can see that iteration is wrong.  

but it gets worse than that because Mapping .keys(), values(), items() don't return lists.  they return KeysView, ValuesView, and ItemsView.  lists are not accepted.


Therefore I don't think it's very feasible we use Mapping here.    

@Dreamsorcerer  ColumnCollection is a public type for this reason.   I would use it directly here for methods that expect a SQLAlchemy ColumnCollection.  Otherwise I'd use a protocol that does just the ``__getitem__()`` part, `Mapping` has too many specifics here.

**Reply by CaselIT** [reply]
Forgot about the difference in iter. I think that breaks the mapping semantics, even disregarding the different types returned by keys, values and items

As a side note, ColumnCollection is probably a good target for a cython version, at least partially
